README.md
Introduction
============

![Logo](/apple-touch-icon-precomposed.png "Logo")

[Où est mon bus ?](http://ouestmonbus.com) est une carte interactive en ligne basée sur OpenStreetMap et les données OpenData de Rennes Métropole.

L'application indique les arrêts de bus en temps réel.

En cliquant sur les arrêts de bus, les prochains passages des lignes sont affichés en direct en précisant un éventuel retard ou une avance.

La position des bus est indiquée sur la carte avec un rafraîchissement
automatique toutes les minutes et une animation donne une approximation
du déplacement des bus.

Caractéristiques principales :
------------------------------

* arrêts de bus affichés aux arrêts en temps réel, si un bus est prévu dans la journée.
* affichage des deux prochains passages de bus en temps réel sur les arrêts.
* affichage de la position des bus sur la carte.
* géolocalisation automatique ou avec saisie d'un lieu référencé sur OpenStreetMap (avec auto-complétion).
* informations traffic.
* informations @starbusmetro en temps réel.
* partage de liens avec localisation.
* consultation des fiches horaires officielles via les icônes des lignes.
* détails de la carte OpenStreetMap complets jusqu'au zoom maximum.

Ouestmonbus.com est une application simple et rapide pour ne plus rater son bus ! 

Elle peut être utilisée sans installation, simplement en se rendant sur le site [Où est mon bus ?](http://ouestmonbus.com) depuis un ordinateur, mobile ou tablette. 

Une version sur le store (Android dans un premier temps, est en cours de préparation).

Fonctionnalités détaillées:
===========================

* Les arrêts de bus sont affichés avec le symbole ![bus](/images/picto/22/station.png "bus") si un bus est prévu avant la fin du la journée.
* Pour obtenir de meilleurs performances d'affichage et une meilleure visibilité, lorsque le zoom est inférieur à 16 sur la carte, les arrêts sont regroupés dans des bulles.
* Les horaires indiqués sont récupérés via les services du STAR et sont en temps réel.
* Les trajets des lignes de l'arrêt sont affichés sur la carte avec leurs couleurs.
* Les icônes de ligne de l'arrêt mettent en premier plan la ligne sur la carte.
* En cliquant sur le trajet sur la carte, le titre apparaît et un lien vers la fiche horaire est disponible.
* Les deux prochains passages dans l'heure sont indiqués.
* Le fond de carte provient d’OpenStreetMap et contient l'ensemble des données associées.
* L'avance ou le retard du bus est indiqué par rapport à l'horaire théorique de passage sur le prochain passage uniquement.
* L'URL (l'adresse) de la page contient les coordonnées, en partageant l'adresse complète vous partagez également la position de la carte:
  * par exemple http://ouestmonbus.com/#map=18/48.10874/-1.67512 partage la position du lycée Emile Zola.
* Le titre de l'arrêt de bus permet de centrer l'arrêt dans la carte.
* Lorsque le site est ouvert à l'emplacement d'un arrêt de bus, les informations de passage s’ouvrent automatiquement
  * par exemple http://ouestmonbus.com/#map=17/48.11274/-1.68021 affiche les passages à l'arrêt Champ Jacquet.
* La présence des arrêts sur la carte sont automatiquement rafraîchie toutes les 3 minutes sans actions utilisateurs.
* Les informations d'une station en cours d'affichage est rafraîchie toutes les minutes.
* Le titre de la page indique le lieu centré sur la carte.
* Intégration avec Twitter.
* Affichage des infos trafic via l'API star.
* Affichage les infos Twitter @starbusmetro.
* Saisie manuelle d'un lieu avec le clavier (service http://photon.komoot.de/)
* Sauvegarde des préférences (bulles d'aide déjà vus et dernière localisation)
* Affichage des bus lors du chargement de la station.
* Bouton zoom pour voir tous les bus en même temps avec la station interrogée.
* Rafraîchissement automatique de la position des bus (1 min) avec une barre de progression.
* Mise en cache des données dans le navigateur (pictogrammes et stations) quand c'est possible.

Fonctionnement technique et développement:
==========================================

Vous voulez modifier, copier, distribuer, [Où est mon bus ?](http://ouestmonbus.com) ?

Vous pouvez, en respectant la [licence GPL](http://www.gnu.org/licenses/gpl-3.0.fr.html) ! ([version complète licence](/COPYING))

Les sources de la version en ligne sont téléchargeables sous forme d'archive [ici](/ouestmonbus.zip). 

Voici quelques explications sur le fonctionnement du logiciel.

Les données sont traitées en deux parties:
* la liste des stations générées avec le script generate.js (nodejs) chaque jour avec une tâche cron.
* les informations des passages (API explore en indiquant le numéro *idarret* / *stop_id* de la station)

La liste des stations est générée à partir des données [Données théoriques du réseau STAR](https://data.keolis-rennes.com/fr/les-donnees/donnees-telechargeables.html).
J'ai essayé de générer la liste des stations avec l'API explore, mais les performances sont moins bonnes, et la précision des informations (localisation de la station) moins juste.

Le format est le suivant:

    {
      "type": "Feature",
      "geometry": {
        "type": "Point",
        "coordinates": [
          "-1.6787340000",
          "48.1094050000"
        ]
      },
      "properties": {
        "id": "1167",
        "name": "République Pré Botté",
        "terminus": [
          "0053,1",
          "0064,0"
        ],
        "limits": [
          "06:25:00",
          "21:21:00"
        ],
        "lines": [
          "0053",
          "0064"
        ]
      }
    }

Pour m'aider au traitement des données j'ai réalisé un schéma au début de mon développement:

![schéma](/docs/desc.png "schéma des données")

Un script télécharge automatiquement les derniers fichiers et génère les données suivantes:

- /data/ -> répertoire de données
- /data/downloads/ -> répertoire des zip téléchargés
- /data/GTFS/2015-08-13/ -> les données GTFS qui commence à partir du 2015-08-13	
- /data/today/ -> données statiques générés chaque jour par ouestmonbus.com
- /data/today/2015-07-19/ -> données statiques générés pour le 2015-07-19
- /data/today/2015-07-19/stations.geojson -> stations utilisés le 2015-07-19
- /data/today/2015-07-19/station_1035.json -> tous les passages prévus le 2015-07-19 à la station 1035 (pas encore implémenté, peut-être dans le futur)

Tuiles:
-------

Les tuiles proviennent d'une génération personnelle à partir des données de OpenStreetMap et Mapnik.
Elles sont hébergées sur le serveur de [Où est mon bus ?](http://ouestmonbus.com).

Tests:
------

La [page tests](/tests/) contient des tests unitaires réalisés avec [Jasmine](http://jasmine.github.io/).

Installer votre propre instance de ouestmonbus.com
==================================================

TODO

Reste à faire pour la version en cours:
=======================================

* notif de nouvelle version
* afficher une notif quand les données de la ligne ne sont pas dispo

Partiellement fait:
-------------------


Idées de développement pour une future version majeure:
=======================================================

* info trafic contextualisé en fonction des lignes sélectionnées [M]

  -> implémenter /api/records/1.0/search?dataset=tco-bus-topologie-lignes-td
  -> correspondance nom court - id
  -> filtrage

* fonction alerte de passage en html5 [XXL à découpé]

Optimisation possible:
----------------------

* utiliser momentjs fromnow pour l'affichage du temps restant

Bugs à corriger:
================

* mode paysage sur petite résolution

Non urgent:
-----------

* rendre robuste le générateur si téléchargement auto des zip (si zip incorrect)

Crédits:
========

© Copyright 2015 - [Benoît Meunier](mailto:ouestmonbus@systemd.info) & Karin Cibard.

![licence GPL](/images/gplv3.png "licence GPL")
[Où est mon bus ?](http://ouestmonbus.com) est un logiciel libre sous [licence GPL](http://www.gnu.org/licenses/gpl-3.0.fr.html).

* Les données des horaires et les icônes du réseau STAR proviennent de https://data.keolis-rennes.com et https://data.explore.star.fr/explore/.
* Les tacets des lignes proviennent de http://www.data.rennes-metropole.fr/.
* Les icônes des bus sont réalisés par Freepik du site http://www.flaticon.com.
* Le reste du design provient de la librairie Metro UI CSS http://metroui.org.ua/ et Leaflet http://leafletjs.com/.



CHANGELOG
* 2015-09-14 - 1.9.6 : Amélioration du temps réel, prise en compte décalage API
* 2015-09-14 - 1.9.5 : CDN sur toutes les ressources
* 2015-09-14 - 1.9.4 : timer de refresh moins rapide
* 2015-09-14 - 1.9.3 : utilisation du CDN
* 2015-09-13 - 1.9.2 : version avec bus qui bougent stable
* 2015-09-12 - 1.9.1 : first test bus qui bouge
* 2015-09-12 - 1.9.0 : les bus bougent first step
* 2015-09-11 - 1.8.6 : publication version avec les lignes
* 2015-09-11 - 1.8.5 : icone groupe arret de bus de même couleur que les arrets
* 2015-09-11 - 1.8.4 : fix hour selection for N2
* 2015-09-11 - 1.8.3 : add leaflet-knn
* 2015-09-10 - 1.8.2 : affichage de toutes lignes dispo à l'arrêt
* 2015-09-10 - 1.8.1 : ajout de sens et de la couleur pour la popup des lignes
* 2015-09-10 - 1.8.0 : Ajout des lignes sur la carte
* 2015-09-09 - 1.7.0 : affichage des lignes dev en cours
* 2015-09-09 - 1.6.11 : publication stable
* 2015-09-09 - 1.6.9 : format du changelog amélioré
* 2015-09-09 - 1.6.8 : ajout de suppression du cache lors du refresh de version
* 2015-09-09 - 1.6.7 : amélioration divers affichage
* 2015-09-09 - 1.6.x : Ajout des stations de métro


src/leaflet-photon.js
L.PhotonBase = L.Class.extend({

    forEach: function (els, callback) {
        Array.prototype.forEach.call(els, callback);
    },

    ajax: function (callback, thisobj) {
        if (typeof this.xhr === 'object') {
            this.xhr.abort();
        }
        this.xhr = new XMLHttpRequest();
        var self = this;
        this.xhr.open('GET', this.options.url + this.buildQueryString(this.getParams()), true);

        this.xhr.onload = function(e) {
            self.fire('ajax:return');
            if (this.status === 200) {
                if (callback) {
                    var raw = this.response;
                    raw = JSON.parse(raw);
                    callback.call(thisobj || this, raw);
                }
            }
            delete this.xhr;
        };

        this.fire('ajax:send');
        this.xhr.send();
    },

    buildQueryString: function (params) {
        var queryString = [];
        for (var key in params) {
            if (params[key]) {
                queryString.push(encodeURIComponent(key) + '=' + encodeURIComponent(params[key]));
            }
        }
        return queryString.join('&');
    },

    featureToPopupContent: function (feature) {
        var container = L.DomUtil.create('div', 'leaflet-photon-popup'),
            title = L.DomUtil.create('h3', '', container);
        title.innerHTML = feature.properties.label;
        return container;
    }

});


L.PhotonSearch = L.PhotonBase.extend({

    includes: L.Mixin.Events,

    options: {
        url: 'http://photon.komoot.de/api/?',
        placeholder: 'Start typing...',
        minChar: 3,
        limit: 5,
        lang: 'fr',
        submitDelay: 300,
        includePosition: true,
        noResultLabel: 'No result',
        feedbackEmail: 'photon@komoot.de',  // Set to null to remove feedback box
        feedbackLabel: 'Feedback'
    },

    CACHE: '',
    RESULTS: [],
    KEYS: {
        LEFT: 37,
        UP: 38,
        RIGHT: 39,
        DOWN: 40,
        TAB: 9,
        RETURN: 13,
        ESC: 27,
        APPLE: 91,
        SHIFT: 16,
        ALT: 17,
        CTRL: 18
    },

    initialize: function (map, input, options) {
        this.map = map;
        this.input = input;
        L.setOptions(this, options);
        var CURRENT = null;

        try {
            Object.defineProperty(this, 'CURRENT', {
                get: function () {
                    return CURRENT;
                },
                set: function (index) {
                    if (typeof index === 'object') {
                        index = this.resultToIndex(index);
                    }
                    CURRENT = index;
                }
            });
        } catch (e) {
            // Hello IE8
        }
        this.input.type = 'search';
        this.input.placeholder = this.options.placeholder;
        this.input.autocomplete = 'off';
        this.input.autocorrect = 'off';
        L.DomEvent.disableClickPropagation(this.input);

        L.DomEvent.on(this.input, 'keydown', this.onKeyDown, this);
        L.DomEvent.on(this.input, 'input', this.onInput, this);
        L.DomEvent.on(this.input, 'blur', this.onBlur, this);
        L.DomEvent.on(this.input, 'focus', this.onFocus, this);
        this.createResultsContainer();
    },

    createResultsContainer: function () {
        this.resultsContainer = L.DomUtil.create('ul', 'photon-autocomplete', document.querySelector('body'));
    },

    resizeContainer: function()
    {
        var l = this.getLeft(this.input);
        var t = this.getTop(this.input) + this.input.offsetHeight;
        this.resultsContainer.style.left = l + 'px';
        this.resultsContainer.style.top = t + 'px';
        var width = this.options.width ? this.options.width : this.input.offsetWidth - 2;
        this.resultsContainer.style.width = width + 'px';
    },

    onKeyDown: function (e) {
        switch (e.keyCode) {
            case this.KEYS.TAB:
                if(this.CURRENT !== null)
                {
                    this.setChoice();
                }
                L.DomEvent.stop(e);
                break;
            case this.KEYS.RETURN:
                L.DomEvent.stop(e);
                this.setChoice();
                break;
            case this.KEYS.ESC:
                L.DomEvent.stop(e);
                this.hide();
                this.input.blur();
                break;
            case this.KEYS.DOWN:
                if(this.RESULTS.length > 0) {
                    if(this.CURRENT !== null && this.CURRENT < this.RESULTS.length - 1) { // what if one resutl?
                        this.CURRENT++;
                        this.highlight();
                    }
                    else if(this.CURRENT === null) {
                        this.CURRENT = 0;
                        this.highlight();
                    }
                }
                break;
            case this.KEYS.UP:
                if(this.CURRENT !== null) {
                    L.DomEvent.stop(e);
                }
                if(this.RESULTS.length > 0) {
                    if(this.CURRENT > 0) {
                        this.CURRENT--;
                        this.highlight();
                    }
                    else if(this.CURRENT === 0) {
                        this.CURRENT = null;
                        this.highlight();
                    }
                }
                break;
        }
    },

    onInput: function (e) {
        if (typeof this.submitDelay === 'number') {
            window.clearTimeout(this.submitDelay);
            delete this.submitDelay;
        }
        this.submitDelay = window.setTimeout(L.Util.bind(this.search, this), this.options.submitDelay);
    },

    onBlur: function (e) {
        this.fire('blur');
        var self = this;
        setTimeout(function () {
            self.hide();
        }, 100);
    },

    onFocus: function (e) {
        this.fire('focus');
        this.input.select();
        this.search();  // In case we have a value from a previous search.
    },

    clear: function () {
        this.RESULTS = [];
        this.CURRENT = null;
        this.CACHE = '';
        this.resultsContainer.innerHTML = '';
    },

    hide: function() {
        this.fire('hide');
        this.clear();
        this.resultsContainer.style.display = 'none';
    },

    setChoice: function (choice) {
        choice = choice || this.RESULTS[this.CURRENT];
        if (choice) {
            this.hide();
            this.fire('selected', {choice: choice.feature});
            this.onSelected(choice.feature);
            this.input.value = '';
        }
    },

    search: function() {
        var val = this.input.value;
        var minChar = typeof this.options.minChar === 'function' ? this.options.minChar(val) : val.length >= this.options.minChar;
        if (!val || !minChar) return this.clear();
        if(val + '' === this.CACHE + '') return;
        else this.CACHE = val;
        this._doSearch();
    },

    _doSearch: function () {
        this.ajax(this.handleResults, this);
    },

    _onSelected: function (feature) {
        this.map.setView([feature.geometry.coordinates[1], feature.geometry.coordinates[0]], this.map.DEFAULT_ZOOM);
        app.localisation.showLocation.call(app.localisation,L.latLng([feature.geometry.coordinates[1], feature.geometry.coordinates[0]]),false,false,1);
        $(".photon-input").hide();
    },

    onSelected: function (choice) {
        return (this.options.onSelected || this._onSelected).call(this, choice);
    },

    _formatResult: function (feature, el) {
        var title = L.DomUtil.create('strong', '', el),
            detailsContainer = L.DomUtil.create('small', '', el),
            details = [],
            type = this.formatType(feature);
        title.innerHTML = feature.properties.name;
        if (type) details.push(type);
        if (feature.properties.city && feature.properties.city !== feature.properties.name) {
            details.push(feature.properties.city);
        }
        if (feature.properties.country) details.push(feature.properties.country);
        detailsContainer.innerHTML = details.join(', ');
    },

    formatResult: function (feature, el) {
        return (this.options.formatResult || this._formatResult).call(this, feature, el);
    },

    formatType: function (feature) {
        return (this.options.formatType || this._formatType).call(this, feature);
    },

    _formatType: function (feature) {
        return feature.properties.osm_value;
    },

    createResult: function (feature) {
        var el = L.DomUtil.create('li', '', this.resultsContainer);
        this.formatResult(feature, el);
        var result = {
            feature: feature,
            el: el
        };
        // Touch handling needed
        L.DomEvent.on(el, 'mouseover', function (e) {
            this.CURRENT = result;
            this.highlight();
        }, this);
        L.DomEvent.on(el, 'mousedown', function (e) {
            this.setChoice();
        }, this);
        return result;
    },

    resultToIndex: function (result) {
        var out = null;
        this.forEach(this.RESULTS, function (item, index) {
            if (item === result) {
                out = index;
                return;
            }
        });
        return out;
    },

    handleResults: function(geojson) {
        var self = this;
        this.clear();
        this.resultsContainer.style.display = 'block';
        this.resizeContainer();
        this.forEach(geojson.features, function (feature) {
            var southWest = L.latLng(47.9690, -1.95019);
            var northEast = L.latLng(48.30274, -1.4649);
            var maxBounds = L.latLngBounds(southWest, northEast);
            if(maxBounds.contains(L.latLng(feature.geometry.coordinates[1],feature.geometry.coordinates[0])))
            {
                self.RESULTS.push(self.createResult(feature));
            }          
        });
        if (geojson.features.length === 0) {
            var noresult = L.DomUtil.create('li', 'photon-no-result', this.resultsContainer);
            noresult.innerHTML = this.options.noResultLabel;
        }
        if (this.options.feedbackEmail) {
            var feedback = L.DomUtil.create('a', 'photon-feedback', this.resultsContainer);
            feedback.href = 'mailto:' + this.options.feedbackEmail;
            feedback.innerHTML = this.options.feedbackLabel;
        }
        this.CURRENT = 0;
        this.highlight();
        if (this.options.resultsHandler) {
            this.options.resultsHandler(geojson);
        }
    },

    highlight: function () {
        var self = this;
        this.forEach(this.RESULTS, function (item, index) {
            if (index === self.CURRENT) {
                L.DomUtil.addClass(item.el, 'on');
            }
            else {
                L.DomUtil.removeClass(item.el, 'on');
            }
        });
    },

    getLeft: function (el) {
        var tmp = el.offsetLeft;
        el = el.offsetParent;
        while(el) {
            tmp += el.offsetLeft;
            el = el.offsetParent;
        }
        return tmp;
    },

    getTop: function (el) {
        var tmp = el.offsetTop;
        el = el.offsetParent;
        while(el) {
            tmp += el.offsetTop;
            el = el.offsetParent;
        }
        return tmp;
    },

    getParams: function () {
        return {
            q: this.CACHE + " Ille-et-Vilaine, Bretagne, France",
            lang: this.options.lang,
            limit: this.options.limit,
            lat: this.options.includePosition ? this.map.getCenter().lat : null,
            lon: this.options.includePosition ? this.map.getCenter().lng : null
        };
    }

});


L.Control.Photon = L.Control.extend({

    includes: L.Mixin.Events,

    onAdd: function (map, options) {
        this.map = map;
        this.container = L.DomUtil.create('div', 'leaflet-photon');

        this.options = L.Util.extend(this.options, options);

        this.input = L.DomUtil.create('input', 'photon-input', this.container);
        this.search = new L.PhotonSearch(map, this.input, this.options);
        this.search.on('blur', this.forwardEvent, this);
        this.search.on('focus', this.forwardEvent, this);
        this.search.on('hide', this.forwardEvent, this);
        this.search.on('selected', this.forwardEvent, this);
        this.search.on('ajax:send', this.forwardEvent, this);
        this.search.on('ajax:return', this.forwardEvent, this);
        return this.container;
    },

    // TODO onRemove

    forwardEvent: function (e) {
        this.fire(e.type, e);
    }


});

L.Map.addInitHook(function () {
    if (this.options.photonControl) {
        this.photonControl = new L.Control.Photon(this.options.photonControlOptions || {});
        this.addControl(this.photonControl);
    }
});

L.PhotonReverse = L.PhotonBase.extend({

    includes: L.Mixin.Events,

    options: {
        url: 'http://photon.komoot.de/reverse/?',
        limit: 1,
        handleResults: null
    },

    initialize: function (options) {
        L.setOptions(this, options);
    },

    doReverse: function (latlng) {
        latlng = L.latLng(latlng);
        this.fire('reverse', {latlng: latlng});
        this.latlng = latlng;
        this.ajax(this.handleResults, this);
    },

    _handleResults: function (data) {
        /*eslint-disable no-console */
        console.log(data);
        /*eslint-enable no-alert */
    },

    handleResults: function (data) {
        return (this.options.handleResults || this._handleResults).call(this, data);
    },

    getParams: function () {
        return {
            lang: this.options.lang,
            limit: this.options.limit,
            lat: this.latlng.lat,
            lon: this.latlng.lng
        };
    }

});
